### ISSUE TITLE
Including a high-dimensional fixed effect 

### ISSUE THREAD

**Issue by hjohns12** [JuliaStats/MixedModels.jl#120]
I want to compare the results from two models: one model with two simple, scalar random effects of the form (1| G) + (1 | H), and one model with a random effect on the first grouping variable and fixed effect on the second ((1 | H) + G). When I specify the second model using the MixedModels package, I’m thrown the “out of memory” error. Because I can't share my actual data, I'm using an example set. In my data, the term G is a high-dimensional categorical variable with 1,400 levels, the term H is a categorical variable with 1.7 million levels, and the total dataset has 6.5 million rows. 

`using DataFrames, RData, MixedModels, NLopt`
`const dat = convert(Dict{Symbol,DataFrame}, load(Pkg.dir("MixedModels", "test", "dat.rda")))`
`m = fit!(lmm(@formula(Y ~ 1 + (1|G) + (1|H)), dat[:ml1m]), true)`

At each evaluation of the objective the first value is the profiled deviance (negative twice the log-likelihood) and the vector is the ratios of the standard deviation of the random effects to the residual standard deviation.  Let's call this θ.  So, 0.399426 = 0.36034996/0.90216887 where the two values on the right hand side come from the variance components table.  The smaller these values are, the more "shrinkage" or "regularization" is being applied to the random effects.  The fixed effects model for G corresponds to θ₂ → ∞ which is no shrinkage.  In practice the optimal value is usually less than 1 and almost never above 10.  So, an effectively infinite value of θ₂ might be 100 or so.

@dmbates suggested working out the constrained optimization so that I can compare the two models. Working it out by hand, he got: 

`opt = NLopt.Opt(:LN_BOBYQA, 1)`
`lower_bounds!(opt, zeros(1))`
`ftol_rel!(opt, 1e-12)`
`ftol_abs!(opt, 1e-8)`
`const θ = fill(1000.0, 2)`

`function obj(x, g)`
`           θ[1] = x[1]`
`           objective(updateL!(setθ!(m, θ)))`
`       end`
`obj([1.], Float64[])  # a check on the function`
`min_objective!(opt, obj)`
`optimize!(opt, ones(1))`

What do folks think of this approach? What's the best way to compare the two models? Thanks in advance!

**Comment by dmbates** [comment]
Thanks for creating the issue, Hope.

As you mentioned in our earlier off-line conversation the central computation in models with fixed effects and random effects is solving a least squares problem.  In the case of mixed-effects models it is a penalized least squares problem where the penalty is determined by the variance structure of the random effects.  Random effects with a low variance have a high penalty on their length; those with a high variance have a smaller penalty.  Fixed-effects terms have no penalty which is both good and bad news.  The good news is that they don't need an iterative optimization of the log-likelihood when evaluating estimates.  The bad news is that without a penalty the model matrix can become singular.

It is a common practice when fitting models with large numbers of parameters to introduce a bit of regularization. Also, from the statistical point of view, there is the bias-variance tradeoff where introducing a bit of bias through the penalty or regularization provides considerably less variability.

(This, by the way, is why I don't understand econometricians wanting to use fixed effects even when the number of levels of the grouping factor is very large.  That is almost never a good choice.)

When you have grouping factors like `G` and `H` that have a very large number of levels, the model matrices will be large and sparse.  In your case the model matrix for the `H` term is 6.5 million by 1.7 million but each row has only one non-zero element.  Essentially it is represented in `MixedModels` by the grouping factor itself (a `ScalarFactorReTerm`)  That part of the model matrix has orthogonal columns so forming X'X results in a great big diagonal block in the upper left.  When forming the Cholesky factor that block stays diagonal.  The next block in the model matrix is 6.5 million by 1400 and also has only one non-zero per row.  In forming X'X this block also forms a diagonal block (of size 1400 by 1400) and a sparse off-diagonal block (the one in the lower triangle is of size 1400 by 1.7 million)  The diagonal is not preserved in the Cholesky factorization but that is not as big a problem because the part that gets "filled in" is only 1400 by 1400.

You can check the nature of the blocks with the `describeblocks` function.
```
julia> const dat = convert(Dict{Symbol,DataFrame}, load(Pkg.dir("MixedModels", "test", "dat.rda")));

julia> m = updateL!(lmm(@formula(Y ~ 1 + (1|G) + (1|H)), dat[:ml1m]));

julia> describeblocks(m)
1,1: Diagonal{Float64} (6040, 6040) Diagonal{Float64}
2,1: SparseMatrixCSC{Float64,Int32} (3706, 6040) SparseMatrixCSC{Float64,Int32}
2,2: Diagonal{Float64} (3706, 3706) Array{Float64,2}
3,1: Array{Float64,2} (1, 6040) Array{Float64,2}
3,2: Array{Float64,2} (1, 3706) Array{Float64,2}
3,3: Array{Float64,2} (1, 1) Array{Float64,2}
4,1: Array{Float64,2} (1, 6040) Array{Float64,2}
4,2: Array{Float64,2} (1, 3706) Array{Float64,2}
4,3: Array{Float64,2} (1, 1) Array{Float64,2}
4,4: Array{Float64,2} (1, 1) Array{Float64,2}
```

I wrote this mostly for myself.  The first matrix type in each row is the block in the X'X type of matrix and the second is the block in the lower Cholesky factor.  Thus 1,1 is Diagonal in both cases but 2,2 is Diagonal in the original but a dense array in the factor.

I'll write more about this later.

**Comment by dmbates** [comment]
What I proposed as a work-around was to tweak the existing structure for a random-effects term to fit the model with fixed-effects for one of the grouping factors.  The central issue is that you can't use the full set of indicators for a grouping factor in addition to the "Intercept" column when working with fixed effects, but you can with random effects because of the shrinkage involved in fitting models with random effects.

I thought it would be possible to adapt the structure for the mixed-effects model by just imposing a very large relative standard deviation on random effects instead of fitting fixed effects but that just masks the singularity problem.  It seems that the approach I had in mind just masks the singularity rather than providing an alternative estimation approach.

So I don't have a good answer now.

**Comment by hjohns12** [comment]
Perhaps it’s too obvious, but my first instinct is that we could drop the intercept term and the model would be identifiable. Why isn't this an option?

**Comment by dmbates** [comment]
In your case it may work to drop the intercept term.  In the example with the MovieLens data you would need to specify the model as
```julia
@formula(Y ~ 0 + (1|G) + (1|H))
```
and the current implementation of ModelMatrix balks at creating a model matrix with zero columns for the fixed effects.

The better solution will be to create a sparse or blocked model matrix for the fixed-effects terms.  I was trying to finesse that by saying that a fixed-effects term behaves like a random-effects term with an artificially large variance of the random effects but that doesn't carry through cleanly.  @kleinschmidt and I and, perhaps, others should look at the [`StatsModels`](https://github.com/JuliaStats/StatsModels.jl) package carefully to see what can be done there.

**Comment by amillscampisi** [comment]
I might have a similar problem. I am trying to estimate a mixed effect model with ~2 million customers. I was able to get a toy version of the model running with 100,000 random customers. At 500,000 individuals (~760k observations), the model crashes with the error below. Is there a simple solution to this problem? 

`fm1 = MM.fit(MM.LinearMixedModel, MM.@formula(spend_centered ~ 0 + (1|cust_key)), df)`

> OutOfMemoryError()
> 
> Stacktrace:
>  [1] Type at ./boot.jl:396 [inlined]
>  [2] Type at ./boot.jl:404 [inlined]
>  [3] zeros at ./array.jl:467 [inlined]
>  [4] Array{Float64,2}(::LinearAlgebra.UniformScaling{Float64}, ::Tuple{Int64,Int64}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/LinearAlgebra/src/uniformscaling.jl:318
>  [5] Type at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/LinearAlgebra/src/uniformscaling.jl:327 [inlined]
>  [6] Type at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/LinearAlgebra/src/uniformscaling.jl:326 [inlined]
>  [7] contrasts_matrix(::StatsModels.DummyCoding, ::Int64, ::Int64) at /home/ec2-user/.julia/packages/StatsModels/AYB2E/src/contrasts.jl:275
>  [8] StatsModels.ContrastsMatrix(::StatsModels.DummyCoding, ::Array{Int64,1}) at /home/ec2-user/.julia/packages/StatsModels/AYB2E/src/contrasts.jl:173
>  [9] evalcontrasts(::DataFrames.DataFrame, ::Dict{Any,Any}) at /home/ec2-user/.julia/packages/StatsModels/AYB2E/src/modelframe.jl:126
>  [10] #ModelFrame#17(::Dict{Any,Any}, ::Type, ::StatsModels.Terms, ::DataFrames.DataFrame) at /home/ec2-user/.julia/packages/StatsModels/AYB2E/src/modelframe.jl:148
>  [11] Type at ./none:0 [inlined]
>  [12] #ModelFrame#20 at /home/ec2-user/.julia/packages/StatsModels/AYB2E/src/modelframe.jl:157 [inlined]
>  [13] Type at ./none:0 [inlined]
>  [14] #LinearMixedModel#12(::Array{Any,1}, ::Dict{Any,Any}, ::Distributions.Normal{Float64}, ::Type, ::StatsModels.Formula, ::DataFrames.DataFrame) at /home/ec2-user/.julia/packages/MixedModels/vY30Z/src/pls.jl:63
>  [15] fit(::Type{MixedModels.LinearMixedModel}, ::StatsModels.Formula, ::DataFrames.DataFrame) at /home/ec2-user/.julia/packages/MixedModels/vY30Z/src/pls.jl:63
>  [16] top-level scope at In[27]:1

**Comment by Nosferican** [comment]
For absorbing fixed effects I usually use the method of alternating projections to get the annihilated matrix. It is really efficient and scales well to high dimensions. I use it to get the within estimator and the code generalizes to partial demeaning which I use for econometrics slang random effects in panel data world. It might be useful here.

**Comment by dmbates** [comment]
Is the reason for the formula being `spend_centered ~ 0 + (1|cust_key)` because you have centered the response and therefore eliminate the "Intercept" or mean?  I would suggest trying it without centering the data (if possible) and changing the formula to `spend ~ 1 + (1|cust_key)`.  I'm not entirely sure if the evaluation of the objective works properly when the model matrix for the fixed effects has zero columns, like here.  I'll experiment.

**Comment by skanskan** [comment]
The only package I've found able to deal with this amount of data and levels is R's glmmTMB.
It would be great to use its methods on Julia.

**Comment by palday** [comment]
@amillscampisi Have you tried running this on a current release? As part of my work on #160, I've simulated relatively large datasets and can do about 2m observations with 25k levels of the blocking variable on a machine with 8 GB of RAM.

@skanskan The methods used by glmmTMB (using automatic differentiation to evaluate the gradient) are fundamentally than the methods used by MixedModels.jl (gradient-free optimization on a sparse representation).

**Comment by palday** [comment]
@amillscampisi  @hjohns12 @skanskan 

The issues revolving around a model with no fixed effects have been resolved (#253), the other issues related to large datasets seem to be one of (a) model misspecification or (b) you need a bigger computer (but see also above comment), so I'm closing this issue.
